src/base/utils/tar.cpp

```diff
@@ -0,0 +1,230 @@
+        bool isVerified() const { return m_verified; }
+        Types type() const { return static_cast<Types>(m_header.at(TAR_FILETYPE_OFFSET)); }
+        int size() const { return octalToDec(Utils::ByteArray::midView(m_header, TAR_FILESIZE_OFFSET, TAR_FILESIZE_SIZE)); }
+
+    private:
+        int octalToDec(const QByteArray &data) const
+        {
+            // only the ASCII characters from '0' to '7' are valid
+            int i = 0;
+            for (; i < data.size(); ++i) { // skip leading invalid bytes
+                char byte = data.at(i);
+                if (byte >= '0' && byte <= '7')
+                    break;
+            }
+
+            int fileSize = 0;
+            for (; i < data.size(); ++i) {
+                char byte = data.at(i);
+                if (byte < '0' || byte > '7') // skip trailing invalid bytes
+                    break;
+
+                fileSize = (8 * fileSize) + (byte - '0');
+            }
+
+            return fileSize;
+        }
+
+        // The various header fields that hold strings are of known max size.
+        // So in the case the string occupies their max size, the string will not
+        // be null terminated. It is pointless to have std::strlen() run through
+        // the whole buffer when we already know the upper limit. Also the buffer
+        // isn't guaranteed to contain a null byte and std::strlen() will have
+        // undefined behavior.
+        int customStrlen(const QByteArray &str) const
+        {
+            int sz = 0;
+            for (; sz < str.size(); ++sz) {
+                if (str.at(sz) == '\0')
+                    break;
+            }
+
+            return sz;
+        }
+
+        // Verified: When the checksum field gets verified
+        // Null: When all bytes are null. Used to indicate end-of-archive (EOF).
+        // Since both actions require running through the whole buffer, combine them to avoid running
+        // twice through the buffer.
+        void verifyAndCheckNull()
+        {
+            int calculatedChecksum = 0;
+            for (int i = 0; i < TAR_CHECKSUM_OFFSET; ++i)
+                calculatedChecksum += static_cast<uchar>(m_header.at(i));
+
+            calculatedChecksum += TAR_CHECKSUM_SIZE * ' '; // 8 ASCII spaces for the checksum field
+
+            for (int i = TAR_FILETYPE_OFFSET; i < m_header.size(); ++i)
+                calculatedChecksum += static_cast<uchar>(m_header.at(i));
+
+            const int checkSum = octalToDec(Utils::ByteArray::midView(m_header, TAR_CHECKSUM_OFFSET, TAR_CHECKSUM_SIZE));
+            m_null = (calculatedChecksum == TAR_CHECKSUM_SIZE * ' ') && (checkSum == 0);
+            m_verified = calculatedChecksum == checkSum;
+        }
+
+        bool m_null = true;
+        bool m_valid = false;
+        bool m_verified = false;
+
+        const QByteArray m_header;
+    };
+}
+
+
+QHash<QString, QByteArray> Utils::Tar::decompress(const QByteArray &data, bool *ok)
```

Comment:
It's better to use exceptions instead of "ok". Then you can provide more error info to the caller side.

Reply:
I took another approach. Have a look.